### PR TITLE
Allow delay configuration for retrier functions.

### PR DESCRIPTION
**Pull request by elyobo**
Delay control is implemented using a user provided function, which
allows any implementation (e.g. Adrian's curves) desired.

The default behaviour is _almost_ maintained; there's an extra
setTimeout with zero delay on each retry, which I don't think is
significant.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -86,38 +86,57 @@ module.exports.createAsyncFnPool = function (fn, concurrency = 1) {
   return Promise.all([...new Array(concurrency)].map(() => enqueue(fn)))
 }
 
+function limitRetrier(limit, delay = 0) {
+  let retries = limit
+  return function () {
```
[the remark this answers is not in the document]
Comment (elyobo) [reply src/index.js]: That is more or less what a generator would be returning, a series of values and then an indicator that it's finished.

**Comment by elyobo** [comment]
I think you're right about generators, it does more or less make sense that way - but my implementation does allow you to use a generator if you happen to want to already, without obliging you to do so.

Re: passing in the attempt instead of the limit... I don't pass in anything to the function at the moment, passing in the last value returned from the function (which seems to be what you're suggesting) might be useful I suppose but it seemed redundant - a closure can be used to maintain that state if a particular function cares about it, but not all functions are going to care about it so passing in anything seemed a bit pointless.

**Comment by elyobo** [comment]
My implementation would _force_ you to wrap your generator inside another function to use it though, so not the nicest API for it.

**Comment by axdg** [comment]
> I don't pass in anything to the function at the moment, passing in the last value returned from the function (which seems to be what you're suggesting) might be useful I suppose but it seemed redundant

No, I'm suggesting passing in the attempt number... ie. "How many attempts have been made to resolve this promise".

On another note, I've also just realised that my implementation doesn't work. The initial generator function would be passed the `limit`... while `yeild` would return `attempt` number back to the generator. It should work like this (exp example):

```js
function* (limit) {
  let attempt = 0
  while (attempt < limit) {
    attempt = yield (attempt * atempt)
  } 
}
```

This is equivalent to this (assuming we wanted 5 retries);

```js
function () {
   yeild* [0, 1, 2, 4, 8]
}
```

**Comment by axdg** [comment]
@elyobo, I am being pedantic though.... I'd like you to consider this API^^ because it seems incredibly logical to me, but I'm happy to merge as is.

**Comment by elyobo** [comment]
> No, I'm suggesting passing in the attempt number... ie. "How many attempts have been made to resolve this promise".

Still seems redundant; if you want to know that you can use a closure to track how many times your function has been called.  Doesn't seem like something the retrier itself needs to bother with.

I agree that the generator function is a nice API as well, so I'm a bit conflicted.  Requiring an iterator seems like it possibly limits things, but is makes it quite nice for most use cases (you could just pass an array of delays and it would also work, for example, no need for the generator function even).  Permitting any function is more flexible, but it makes using generators less pleasant.

**Comment by axdg** [comment]
Por que no los dos? We can make both APIs work, by checking if the function that is passed *is* a generator function or a regular function. If it's not a generator function, we can wrap it to make an iterable out of it's return values.

**Comment by elyobo** [comment]
Summary of Slack discussions around this: it would be nicer if the second argument could be more flexible (number, function, generator), and if a max number of retries could also be provided to make providing infinite generators (like many curves) easier.

**Pull request merged**